Container Provider Refresh Refactor Image Registries Collection

enoodle opened the pull request:
After collecting images from Openshift new image registries are added
and should be collected. The previous registries are also cleared to
prevent duplicates.

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1418590

enoodle commented:
cc @zeari

enoodle commented:
@miq-bot add_label bug, providers/containers

enoodle commented:
an alternative solution could be to move the collection of openshift images to happen before the kubernetes parsing ( in `super` ).

simon3z commented:
@enoodle do we have a bug for this and clear steps to know what was the problem, how to reproduce, etc.? I am sure we'll be asked all these info anyway.

enoodle commented:
@simon3z I opened a BZ and added to the description

simon3z commented:
@zeari can you review?
@enoodle @zeari we want to try and merge this before Monday 6th.

chrispy1 commented:
@miq-bot add_label euwe/yes

enoodle commented:
I refactored the solution to collect container image registries the moment that we identify them instead of later in a collection function.
Also added tests.

simon3z commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -692,6 +686,8 @@ def parse_container_image(image, imageID)
           @data_index.store_path(
             :container_image_registry, :by_host_and_port, host_port, container_image_registry)
+          @data[:container_image_registries] ||= []
+          @data[:container_image_registries] << container_image_registry
```
@Fryguy @blomquisg this LGTM but I don't know if there's any implication on not using `process_collection` for this data (should be OK for as much as I know).

enoodle replied:
@simon3z an alternative can be using:
```
process_collection_item(container_image_registry, :container_image_registries) { |r| r }
```

simon3z replied:
@enoodle it depends on where you'll put that line. Can you update the PR so we'll see how it looks like.

enoodle replied:
@simon3z I palced that line and move the initialization to the `initialize` function. PTAL

simon3z commented:
LGTM :+1: 
cc @Fryguy @blomquisg 

@miq-bot assign Fryguy

simon3z commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -692,6 +686,8 @@ def parse_container_image(image, imageID)
         if stored_container_image_registry.nil?
           @data_index.store_path(
             :container_image_registry, :by_host_and_port, host_port, container_image_registry)
+          @data[:container_image_registries] ||= []
```
@enoodle maybe we can move this initialization somewhere else?

There's an edge-case where all the images have no registry where this may remain `nil`, right? (As we'll never hit this code).

enoodle replied:
we can move this to `initialize`

simon3z commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -9,6 +9,7 @@ def self.ems_inv_to_hashes(inventory)
 
     def initialize
       @data = {}
+      @data[:container_image_registries] ||= []
```
@enoodle for sure no need for the `||` you just defined it above. Maybe you can even do:

```ruby
@data = {
  # >>> Here you need a meaningful comment that explains this exception <<<
  :container_image_registries => []
}
```

Maybe you can even add a test... a refresh with no registries would it fail if it is `nil` instead of `[]` (if not we don't even need to initialize).

enoodle replied:
@simon3z I tested this and there is no need to initialize this in case no registries are found. 
In case no registries are found then `@data[:container_image_registries]` will return nil, and then `hashes[:container_image_registries]` in [1] will be nil and in [2] we will return without a failure.

[1]https://github.com/ManageIQ/manageiq/blob/master/app/models/ems_refresh/save_inventory_container.rb#L13
[2]https://github.com/ManageIQ/manageiq/blob/master/app/models/ems_refresh/save_inventory_container.rb#L325

simon3z commented on spec/models/manageiq/providers/openshift/container_manager/refresh_parser_spec.rb:
```diff
@@ -106,6 +106,32 @@
+        expect(parser.instance_variable_get('@data')[:container_image_registries].size).to eq(1)
+      end
+
+      it "collects image registries from openshift images" do
```
@enoodle what is this testing? Maybe I am missing something or the description can be improved.

enoodle replied:
@simon3z it uses the `:after` clause to test, this makes sure that if an image registry is defined only from an image found from Openshift, this image registry will be collected.

simon3z replied:
@enoodle then it should probably something more explanatory than `"collects image registries from openshift images"` which is not fully reflecting what you're testing here.

simon3z commented:
@enoodle much better now! Really nice!

enoodle commented:
I changed the spec name to "collects image registries from openshift images that are not also running pods images", also squashed the commits. @simon3z PTAL

enoodle commented:
ping @simon3z @Fryguy

simon3z commented:
@enoodle have you verified that this works also without any registry? (`container_image_registries` with values `nil` or '[]`)

enoodle commented:
@simon3z Yes, I run it while manually disabling image/pod collection disabled and saw that in `save_inventory_container` we get to `save_container_image_registries_inventory` with empty hash ( `hashes == nil` and continues silently without failing.

simon3z commented:
@enoodle OK thanks.

Then waiting for (review/)merge from @Fryguy @chessbyte :+1:

Fryguy commented on spec/models/manageiq/providers/openshift/container_manager/refresh_parser_spec.rb:
```diff
@@ -166,6 +166,32 @@
     end
+
+    context "image registries from openshift images" do
+      after :each do
```
`:each` here is redundant and not needed.

Fryguy commented on spec/models/manageiq/providers/openshift/container_manager/refresh_parser_spec.rb:
```diff
@@ -166,6 +166,32 @@
+
+    context "image registries from openshift images" do
+      after :each do
+        inventory = {"image" => [image_from_openshift,]}
```
Is that trailing comma expected?  It doesn't do anything, so I think it should be removed

enoodle replied:
fixed

Fryguy commented on spec/models/manageiq/providers/openshift/container_manager/refresh_parser_spec.rb:
```diff
@@ -166,6 +166,32 @@
+
+        parser.get_openshift_images(inventory)
+        expect(parser.instance_variable_get('@data_index')[:container_image_registry][:by_host_and_port].size).to eq(1)
+        expect(parser.instance_variable_get('@data')[:container_image_registries].size).to eq(1)
```
Doing expectations in an `after` is very strange, as `after` is meant for cleanup/teardown, and it especially makes the `it` on line 179 very strange as it's empty.  If you want to bunch these together to DRY it up, prefer a regular Ruby method that is called from each `it`, or perhaps even better, use RSpec `shared_examples`

enoodle replied:
I changed to a single function `parse_single_openshift_image_with_registry` as shared_examples seems like something to share between different objects and less between different functions.

enoodle commented:
@Fryguy I made the changes, PTAL

simaishi commented:
Euwe backport details:

```
$ git log -1
commit 6e5ce3176ccb9c5eba7bd0aac0fb43b202866935
Author: Jason Frey <fryguy9@gmail.com>
Date:   Mon Feb 6 13:04:34 2017 -0500

    Merge pull request #13724 from enoodle/openshift_refresh_recollect_image_registries
    
    Container Provider Refresh Refactor Image Registries Collection
    (cherry picked from commit 82b0db17984edf9e019fb0d644c91d0150bfe34a)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1419680
```
